bpf/bpf_lxc.c

```diff
@@ -1941,6 +1941,119 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
 	return ret;
 }
 
+static __always_inline bool
+ipv4_to_endpoint_is_hairpin_flow(struct __ctx_buff *ctx, struct iphdr *ip4, __u32 __maybe_unused src_sec_identity)
+{
+	struct ipv4_ct_tuple tuple = {};
+	struct lb4_backend *backend __maybe_unused;
+	struct ct_entry *entry;
+	struct ct_map *map;
+	int err, l4_off;
+	__be32 pod_ip, service_ip;
+	__be16 client_port, backend_port, service_port;
+
+	/* For reasons unknown, hairpinned packets don't have their IDs resolved and
+	 * are always seen as originating from world.
+	 */
+	if (src_sec_identity != WORLD_ID)
+		return false;
+
+	/* Extract the tuple from the packet so we can freely access addrs and ports.
+	 * All values are in network byte order.
+	 */
+	err = lb4_extract_tuple(ctx, ip4, ETH_HLEN, &l4_off, &tuple);
+	if (IS_ERR(err))
+		return false;
+
+	/* Two different types of packets need to be handled here. client->server and
+	 * server->client packets both pass through the to-endpoint path; presumably
+	 * they are NATed differently depending on the direction. We can observe both
+	 * service_ip:client_port -> pod_ip:service_port and
+	 * IPV4_LOOPBACK:backend_port -> pod_ip:client_port.
+	 */
+	if (tuple.saddr != IPV4_LOOPBACK) {
+		/* 
+		 * If the packet originates from a non-loopback address, it will look like:
+		 * 10.96.211.200:12345 -> 10.244.1.101:8080 or service_ip:client_port ->
+		 * pod_ip:service_port.
+		 *
+		 * In order to determine whether the packet is a hairpinned packet, we need to
+		 * obtain the backend (listen) port first, requiring a CT lookup with the
+		 * TUPLE_F_SERVICE flag, followed by a backend lookup. After this, the regular
+		 * CT TUPLE_F_OUT lookup can proceed.
+		 *
+		 * We're looking for the following CT entry:
+		 * daddr 10.96.211.200 (service ip)
+		 * saddr 10.244.1.101 (pod ip)
+		 * dport 12345 (client port)
+		 * sport 8080 (service port)
+		 * nexthdr 06 (TCP)
+		 * flags 04 (TUPLE_F_SERVICE)
```

Comment:
We can maybe drop this part from the comments. The numeric values don't tell much, given that we haven't fully introduced the setup and explained which IP goes to which entity; and your assignments below with the dedicated variables make it very explicit how we build the tuple for the lookup.